yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -616,17 +620,29 @@ private[spark] class Client(
    * throw an appropriate SparkException.
    */
   def run(): Unit = {
-    val (yarnApplicationState, finalApplicationStatus) = monitorApplication(submitApplication())
-    if (yarnApplicationState == YarnApplicationState.FAILED ||
-      finalApplicationStatus == FinalApplicationStatus.FAILED) {
-      throw new SparkException("Application finished with failed status")
-    }
-    if (yarnApplicationState == YarnApplicationState.KILLED ||
-      finalApplicationStatus == FinalApplicationStatus.KILLED) {
-      throw new SparkException("Application is killed")
-    }
-    if (finalApplicationStatus == FinalApplicationStatus.UNDEFINED) {
-      throw new SparkException("The final status of application is undefined")
+    val appId = submitApplication()
+    if (fireAndForget) {
+      logInfo("... polling ResourceManager for application state")
```

Comment:
This isn't actually polling. It's just requesting the application report once and logs it, which isn't particularly useful for the user because s/he will have no idea whether the application has even been successfully launched.

A better way would be to do what standalone mode does this, which is keep requesting the status until it is running or has failed. I believe we can already do this with `monitorApplication(appId, returnOnRunning = true)`.